lua_modules/sntp/sntp.lua

```diff
@@ -0,0 +1,290 @@
+local MAX_SERVER_ATTEMPTS = 2
+local SNTP_TIMEOUT        = 5000
+
+-- Constructor
+--   sc and fc are our Success and Failure Callbacks, resp.
+local new = function(serv, sc, fc, now)
+
+  if     type(serv) == "string" then serv = {serv}
+  elseif serv == nil            then serv =
+    {
+      nil,
+      "1.nodemcu.pool.ntp.org",
+      "2.nodemcu.pool.ntp.org",
+    }
+    local ni = net.ifinfo(0)
+    ni = ni and ni.dhcp
+    serv[1] = ni.ntp_server or "0.nodemcu.pool.ntp.org"
+  elseif type(serv) ~= "table"  then error "Bad server table"
+  end
+
+  if type(sc) ~= "function" and type(sc) ~= "lightfunction" then
+    error "Bad success callback type"
+  end
+  if fc ~= nil and type(fc) ~= "function" and type(fc) ~= "lightfunction" then
+    error "Bad failure callback type"
+  end
+  if now ~= nil and type(now) ~= "function" and type(now) ~= "lightfunction" then
+    error "Bad clock type"
+  end
+  now = now or (rtctime and rtctime.get)
+  if now == nil then error "Need clock function" end
+
+  local self = {}
+  self.servers = serv
+
+  local _tmr -- contains the currently running timer, if any
+  local _udp -- the socket we're using to talk to the world
+
+  local _kod = {} -- kiss of death flags accumulated accoss syncs
+  local _pbest -- best server from prior pass
+
+  local _res -- the best result we've got so far this pass
+  local _best -- best server this pass, for updating _pbest
+
+  local _six -- index of the server in serv to whom we are speaking
+  local _sat -- number of times we've tried to reach this server
+
+  -- Shut down the state machine
+  --
+  -- upvals: _tmr, _udp, _six, _sat, _res, _best
+  local function _stop()
```

Comment:
The destructor in general the cleanup needed.  Also list assignments are useful.  This routine body could have been encoded
```Lua
    _tmr = _tmr and _tmr:unregister()
    _udp = _udp and _udp:close()
    _six,_sat,_res, _best = nil
    for k,v in pairs(_kod) do _kod[k] = (v > 0) and (v - 1) or nil end
```

Reply:
> `_udp:close()` does not unregister callbacks

No but _udp:__gc() does and as `_udp:close()` returns `nil` this clears derefences the socket which triggers the proper GC including clearing down the registry entries.

But there is really a separate `net` tidy up issue that closing a socket should automatically unref any CBs in it.

Reply:
Right, but we won't get to `_udp:__gc()` if there are knots tied through the registry.  Again, I don't mean for there to be, here, but it seemed generally like a belt&suspenders kind of approach.

And yes, as per #3062, I think `net` needs some tidying.  If you think the contract is that `:close()` should rip all callbacks out, it would be good to a note to that effect that there.  I have generally shied away from that particular option because the object outlives `:close()` and can, in principle, be pressed back into service.  At least, the TCP objects can; UDP is a little harder to justify.